kratos/processes/find_intersected_geometrical_objects_process.cpp

```diff
@@ -25,231 +27,457 @@
 
+    const auto it_cond_begin = mrModelPart1.ConditionsBegin();
+
+    #pragma omp parallel for private(leaves)
+    for (int i = 0; i < number_of_conditions; i++) {
+        auto it_cond = it_cond_begin + i;
+        leaves.clear();
+        IdentifyNearEntitiesAndCheckEntityForIntersection(*(it_cond.base()), leaves);
     }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
 
-    void FindIntersectedGeometricalObjectsProcess::GenerateOctree() {
-        this->SetOctreeBoundingBox();
+template<class TEntity>
+void FindIntersectedGeometricalObjectsProcess<TEntity>::ExecuteInitialize()
+{
+    GenerateOctree();
+}
 
-        // Adding mrModelPart2 to the octree
-        for (auto i_node = mrModelPart2.NodesBegin(); i_node != mrModelPart2.NodesEnd(); i_node++) {
-#ifdef KRATOS_USE_AMATRIX   // This macro definition is for the migration period and to be removed afterward please do not use it 
-            mOctree.Insert(i_node->Coordinates().data());
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<>
+std::size_t FindIntersectedGeometricalObjectsProcess<Element>::WorkingSpaceDimension()
+{
+    return mrModelPart1.Elements().begin()->GetGeometry().WorkingSpaceDimension();
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<>
+std::size_t FindIntersectedGeometricalObjectsProcess<Condition>::WorkingSpaceDimension()
+{
+    return mrModelPart1.Conditions().begin()->GetGeometry().WorkingSpaceDimension();
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<>
+void FindIntersectedGeometricalObjectsProcess<Element>::GenerateOctree()
+{
+    this->SetOctreeBoundingBox();
+
+    // Adding mrModelPart2 to the octree
+    for (auto it_node = mrModelPart2.NodesBegin(); it_node != mrModelPart2.NodesEnd(); it_node++) {
+#ifdef KRATOS_USE_AMATRIX   // This macro definition is for the migration period and to be removed afterward please do not use it
+        mOctree.Insert(it_node->Coordinates().data());
 
 #else
-            mOctree.Insert(i_node->Coordinates().data().data());
+        mOctree.Insert(it_node->Coordinates().data().data());
 #endif // ifdef KRATOS_USE_AMATRIX
-        }
+    }
 
-        for (auto i_element = mrModelPart2.ElementsBegin(); i_element != mrModelPart2.ElementsEnd(); i_element++) {
-            mOctree.Insert(*(i_element).base());
-        }
+    // Iterate over the elements
+    for (auto it_elem = mrModelPart2.ElementsBegin(); it_elem != mrModelPart2.ElementsEnd(); it_elem++) {
```

Comment:
Here we do the same but for the other model part, so we can create an specialization for each one of the model parts.